### PR TITLE
Issue #230 Update Docker version in boot2docker ISO

### LINKED ISSUE #230: Update Docker version in boot2docker ISO
Hi,

Currently, the boot2docker image version used to generate [`boot2docker.iso` is `1.12.3`](https://github.com/minishift/minishift/blob/master/deploy/iso/Dockerfile#L17).

However, Docker version is still lower than it i.e [`1.11.2`](https://github.com/minishift/minishift/blob/master/deploy/iso/Dockerfile#L20).

Upstream [boot2docker repo](https://github.com/boot2docker/boot2docker/releases/tag/v1.12.3) also maintain the same version for both.
One definite reason would be "less confusion".

Unless we have any issues with docker version `1.12.3`, we should update it.

Moreover, I found interesting facts today as:
```
#       Boot2docker version 1.12.3 boot2docker.iso 
Boot2Docker version 1.12.3, build HEAD : 7fc7575 - Thu Oct 27 17:23:17 UTC 2016
Docker version 1.12.3, build 6b644ec

#       Minishift master boot2docker.iso
Boot2Docker version 1.11.2, build HEAD : 7fc7575 - Thu Oct 27 17:23:17 UTC 2016
Docker version 1.11.2, build b9f10c9
```
Clearly, minishift master boot2docker.iso is showing wrong info of Boot2docker version for `build HEAD 7fc7575`.

### PR DESCRIPTION
**Pull request by budhrg**
Fix #230 .

### CODE REVIEW

File: deploy/iso/Dockerfile [hunk 1]
```diff
@@ -17,7 +17,7 @@
 FROM boot2docker/boot2docker:1.12.3
 
 # Set the version of Docker and the expected sha.
-RUN echo "1.11.2" > $ROOTFS/etc/version
+RUN echo "1.11.3" > $ROOTFS/etc/version
```
Comment (LalatenduMohanty) [inline deploy/iso/Dockerfile]: @budhrg shouldn't it be 1.12.3 instead of 1.11.3? Also can you please open an issue for it?

Comment (budhrg) [inline deploy/iso/Dockerfile]: Ya @LalatenduMohanty :smile: Updated now

**Comment by praveenkumar** [comment]
LGTM :+1: 

@LalatenduMohanty looks like change which you requested is now part of this patch.

**Comment by LalatenduMohanty** [comment]
@praveenkumar @budhrg Manually updating the same variable in two places is not a good idea IMO. Can we use a variable here , so that when we update the version variable in gets updated in required places?

**Comment by budhrg** [comment]
> @praveenkumar @budhrg Manually updating the same value in two places is not a good idea IMO. Can we use a variable here , so that when we update the version variable it gets updated in required places?

Agreed :+1:

**Comment by budhrg** [comment]
@LalatenduMohanty But it looks like `FROM` must be the [first statement in Dockerfile](https://docs.docker.com/engine/reference/builder/#from). Hence, storing anything before make the Dockerfile invalid.

**Comment by praveenkumar** [comment]
@budhrg @LalatenduMohanty why don't we put something as build args in the build.sh file because there it create the iso, docker support build time args.

build.sh file: https://github.com/minishift/minishift/blob/master/deploy/iso/build.sh

https://docs.docker.com/engine/reference/commandline/build/#/set-build-time-variables---build-arg

**Comment by praveenkumar** [comment]
@budhrg alright so looks like currently it's not possible https://github.com/docker/docker/issues/18119 and https://github.com/faradayio/cage/issues/56 is rfc for same. I can now think of using shell power and do some magic like https://github.com/minishift/minishift-centos-iso/blob/master/Makefile#L36

**Comment by budhrg** [comment]
> I can now think of using shell power and do some magic like https://github.com/minishift/minishift-centos-iso/blob/master/Makefile#L36

That's looks bit complex and I am letting it like that.

**Comment by jimmidyson** [comment]
You could pipe the Dockerfile to stdin of docker build perhaps?

**Comment by praveenkumar** [comment]
@jimmidyson Good idea 👍 , @budhrg try that I think that will work.

**Comment by hferentschik** [comment]
Can we track this stuff in an issue please.

**Comment by hferentschik** [comment]
guys, what makes you believe that this is the same things and that it should be aligned? In the FROM clause we are dealing with the boot2docker/boot2docker image version, in the second case we are dealing with the Docker version to be installed. Why do you think that this should be the same. And even if, looking at the rest of the _Dockerfile_, just updating the docker version without the SHA will most likely fail, right? 

So I am still wondering what this is about. Are we addressing a concrete bug? The pull request title says "Fix boot2docker pinned version" which seems wrong? Is the aim to update the Docker daemon version?

**Comment by praveenkumar** [comment]
@hferentschik FROM section tag represent which boot2docker base image we are going to consume and `$ROOTFS/etc/version` represent which docker version we want in our iso and both are completely different and shouldn't be hard requirement to match as you said. In past openshift had issue like it only work with specific version of docker and this allow us to chose which version we should embedded in our iso. 

@budhrg Please change your commit message and issue tracker details also if that patch will go in, it also need update on SHA and this need test after build that iso before we merge.

**Comment by LalatenduMohanty** [comment]
@hferentschik You are right , I had missed https://github.com/minishift/minishift/blob/master/deploy/iso/Dockerfile#L35-L40 and I was under an impression that docker version ind boot2docker versions stays same as per https://github.com/boot2docker/boot2docker/blob/master/Dockerfile#L288-L292

**Comment by LalatenduMohanty** [comment]
@budhrg I am not sure if we want to keep the docker version same as boot2docker version unless you have a strong reason for this.

**Comment by budhrg** [comment]
Right now I am building iso from Dockerfile with suggested changes to see any regression.

Some changes highlighted here around [docker version 1.12.3](https://github.com/docker/docker/blob/master/CHANGELOG.md#1123-2016-10-26).

**Comment by budhrg** [comment]
Tested the ISO generated with the suggested changes and `minishift start` working fine with proper boot2docker version and docker version.
```
Boot2Docker version 1.12.3, build HEAD : 7fc7575 - Thu Oct 27 17:23:17 UTC 2016
Docker version 1.12.3, build 6b644ec
```

Full logs - https://gist.github.com/budhrg/2858071ec621271b9084622a99a8ad9f

**Comment by hferentschik** [comment]
I am not opposed to upgrading docker provided the ISO build works now and there are not other regressions when it comes to provisioning OpenShift. What I don't understand it why we look into this atm where we for example have also issue #134 which would initially be more valuable. If we had the boot2docker ISO generated extracted, we could add CentOS CI for testing the ISO build and run smoke tests for OpenShift provisioning. This would have caught the issue here from the start.

**Comment by budhrg** [comment]
@hferentschik We can add this change while migrating to new repo. I am ok to wait for that .

**Comment by hferentschik** [comment]
Closing this issue in favor of minishift/minishift-b2d-iso#6. Once minishift-b2d-iso is working we can add a pull request there as a first functional change. Then we can cut a release we can use to test integration with minishift.

**Pull request closed without merging**